main.cpp

```diff
@@ -270,101 +219,165 @@ struct HeapA
     }
 
-    DoubleType& operator+=( double rhs );
-    DoubleType& operator-=( double rhs );
-    DoubleType& operator*=( double rhs );
-    DoubleType& operator/=( double rhs );
-
-    operator double() const { return *value; }
-
-    DoubleType& pow(double d); 
-    DoubleType& pow(const FloatType&);
-    DoubleType& pow(const DoubleType&);
-    DoubleType& pow(const IntType&);
+    Numeric& operator+=( Type rhs )
+    {
+        *value += rhs; 
+        return *this;
+    }
+    Numeric& operator-=( Type rhs )
+    {
+        *value -= rhs; 
+        return *this;
+    }
+    Numeric& operator*=( Type rhs )
+    {
+        *value *= rhs; 
+        return *this;
+    }
 
-    DoubleType& apply(std::function<DoubleType&(double&)>);
-    DoubleType& apply(void(*)(double&));
+    Numeric& operator/=( Type rhs )
+    {
+        if(rhs == 0.0)
+        {
+            std::cout << "warning: floating point division by zero!" << std::endl; 
+        }
 
-private:
-    double* value = nullptr;
-    DoubleType& powInternal(const double d);
+        *value /= rhs;
+        return *this;
+    }
 
-};
+    operator Type() const { return *value; }
 
-struct IntType
-{
-    explicit IntType(int val) : value(new int(val)) {}
-    ~IntType()
+    Numeric& pow(const Type& t)
     {
-        delete value;
+        return powInternal(t);
     }
 
-    IntType& operator+=( int rhs );
-    IntType& operator-=( int rhs );
-    IntType& operator*=( int rhs );
-    IntType& operator/=( int rhs );
-    operator int() const { return *value; }
-    IntType& pow(int i); 
-    IntType& pow(const FloatType&);
-    IntType& pow(const DoubleType&);
-    IntType& pow(const IntType&);
-
-    IntType& apply(std::function<IntType&(int&)>);
-    IntType& apply(void(*)(int&));
+    template<typename T>
+    Numeric& apply(T freeFunc)
+    {
+        freeFunc(*value);
+        return *this;
+    }
 
 private:
-    int* value = nullptr;
-    IntType& powInternal(const int i);
+    std::unique_ptr<Type> value;
+    Numeric& powInternal(const Type& d)
+    {
+        *value = static_cast<Type>(std::pow( *value, d ));
+        return *this;
+    }
 };
 
 struct Point
 {
 
     Point(float m_x, float m_y) : x(m_x), y(m_y) {}
 
-    Point(const FloatType& ft_x, const FloatType& ft_y) : Point(static_cast<float>(ft_x), static_cast<float>(ft_y)) {}
-    Point(const IntType& it_x, const IntType& it_y) : Point(static_cast<float>(it_x), static_cast<float>(it_y)) {}
-    Point(const DoubleType& dt_x, const DoubleType& dt_y) : Point(static_cast<float>(dt_x), static_cast<float>(dt_y)) {}
+    Point(const Numeric<float>& ft_x, const Numeric<float>& ft_y) : Point(static_cast<float>(ft_x), static_cast<float>(ft_y)) {}
+    Point(const Numeric<int>& it_x, const Numeric<int>& it_y) : Point(static_cast<float>(it_x), static_cast<float>(it_y)) {}
+    Point(const Numeric<double>& dt_x, const Numeric<double>& dt_y) : Point(static_cast<float>(dt_x), static_cast<float>(dt_y)) {}
```

Comment:
you can eliminate these constructors if you template the main constructor, and use casting to initialize the member variables, just like you're doing in these 3 constructors. 
- [ ] revise